catalog/core/catalog-core-ftp/src/main/java/ddf/catalog/core/ftp/ftplets/CustomFtplet.java

```diff
@@ -0,0 +1,368 @@
+        this.mimeTypeMapper = mimeTypeMapper;
+    }
+
+    @Override
+    public FtpletResult onConnect(FtpSession session) throws FtpException, IOException {
+        return FtpletResult.DEFAULT;
+    }
+
+    @Override
+    public FtpletResult onDisconnect(FtpSession session) throws FtpException, IOException {
+        return FtpletResult.DEFAULT;
+    }
+
+    @Override
+    public FtpletResult onLogin(FtpSession session, FtpRequest request)
+            throws FtpException, IOException {
+        if (!StringUtils.isNotEmpty(request.getArgument())) {
+            session.write(new DefaultFtpReply(FtpReply.REPLY_451_REQUESTED_ACTION_ABORTED,
+                    "Password cannot be empty"));
+            return FtpletResult.DISCONNECT;
+        }
+
+        AuthenticationToken authenticationToken = new UPAuthenticationToken(session.getUser()
+                .getName(), request.getArgument());
+        try {
+            Subject shiroSubject = securityManager.getSubject(authenticationToken);
+            if (shiroSubject != null) {
+                session.setAttribute(SUBJECT, shiroSubject);
+                return FtpletResult.SKIP;
+            }
+        } catch (SecurityServiceException e) {
+            LOGGER.error("Failure to authenticate subject.", e);
+        }
+        session.write(new DefaultFtpReply(FtpReply.REPLY_550_REQUESTED_ACTION_NOT_TAKEN,
+                "Failed to acquire authenticated subject."));
+        return FtpletResult.DISCONNECT;
+    }
+
+    /**
+     * @param session The current {@link FtpSession}
+     * @param request The current {@link FtpRequest}
+     * @return {@link FtpletResult#SKIP} - signals successful ingest and to discontinue and further processing on the FTP request
+     * @throws FtpException general exception for Ftplets
+     * @throws IOException  thrown when there is an error fetching data from client
+     */
+    @Override
+    public FtpletResult onUploadStart(FtpSession session, FtpRequest request)
+            throws FtpException, IOException {
+
+        LOGGER.debug("Beginning FTP ingest of {}", request.getArgument());
+
+        Subject shiroSubject = (Subject) session.getAttribute(SUBJECT);
+        if (shiroSubject != null) {
+            FtpFile ftpFile = null;
+            String fileName = request.getArgument();
+
+            try {
+                ftpFile = session.getFileSystemView()
+                        .getFile(fileName);
+            } catch (FtpException e) {
+                LOGGER.error("Failed to retrieve file from FTP session");
+            }
+
+            if (ftpFile == null) {
+                LOGGER.error(
+                        "Sending FTP status code 501 to client - syntax errors in request parameters");
+                session.write(new DefaultFtpReply(FtpReply.REPLY_501_SYNTAX_ERROR_IN_PARAMETERS_OR_ARGUMENTS,
+                        "STOR"));
+
+                throw new NullPointerException("File to be transferred from client did not exist");
```

Comment:
NPE's aren't usually good to throw since they typically indicate an error in your code vs an error in operations.  I would recommend throwing one of the typed exceptions (like FtpExcetion from above - actually this whole block should probably be moved into that catch block and simply stop execution there).